prov/gni/src/gnix_sep.c

```diff
@@ -0,0 +1,860 @@
+		if (domain_priv != av->domain) {
+			return -FI_EINVAL;
+		}
+
+		/* We currently only support FI_AV_MAP */
+		if (av->type != FI_AV_MAP) {
+			return -FI_EINVAL;
+		}
+
+		for (i = 0; i < sep->info->ep_attr->tx_ctx_cnt; i++) {
+			ep = container_of(sep->tx_ep_table[i],
+					  struct gnix_fid_ep, ep_fid);
+			if (ep == NULL) {
+				return -FI_EINVAL;
+			}
+			ep->av = av;
+			_gnix_ep_init_vc(ep, av->type);
+			_gnix_ref_get(ep->av);
+		}
+
+		for (i = 0; i < sep->info->ep_attr->rx_ctx_cnt; i++) {
+			ep = container_of(sep->rx_ep_table[i],
+					  struct gnix_fid_ep, ep_fid);
+			if (ep == NULL) {
+				return -FI_EINVAL;
+			}
+			ep->av = av;
+			_gnix_ep_init_vc(ep, av->type);
+			_gnix_ref_get(ep->av);
+		}
+
+		break;
+	default:
+		ret = -FI_ENOSYS;
+		break;
+	}
+
+	return ret;
+}
+
+/*******************************************************************************
+ * Base SEP API function implementations.
+ ******************************************************************************/
+static int gnix_sep_control(fid_t fid, int command, void *arg)
+{
+	int ret = FI_SUCCESS;
+	struct gnix_fid_ep *ep;
+	struct gnix_fid_trx *trx_priv;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	switch (fid->fclass) {
+	case FI_CLASS_SEP:
+		/* nothing to do for scalable endpoints */
+		return FI_SUCCESS;
+	case FI_CLASS_TX_CTX:
+	case FI_CLASS_RX_CTX:
+		trx_priv = container_of(fid, struct gnix_fid_trx, ep_fid);
+		ep = trx_priv->ep;
+		break;
+	default:
+		return -FI_EINVAL;
+	}
+
+	if (!ep) {
+		return -FI_EINVAL;
+	}
+
+	switch (command) {
+	case FI_ENABLE:
+		if (GNIX_EP_RDM_DGM(ep->type)) {
+			if (ep->cm_nic == NULL) {
+				ret = -FI_EOPBADSTATE;
+				goto err;
+			}
+			ret = _gnix_vc_cm_init(ep->cm_nic);
+			if (ret != FI_SUCCESS) {
+				GNIX_WARN(FI_LOG_EP_CTRL,
+				     "_gnix_vc_cm_nic_init call returned %d\n",
+					ret);
+				goto err;
+			}
+			ret = _gnix_cm_nic_enable(ep->cm_nic);
+			if (ret != FI_SUCCESS) {
+				GNIX_WARN(FI_LOG_EP_CTRL,
+				     "_gnix_cm_nic_enable call returned %d\n",
+					ret);
+				goto err;
+			}
+			if (ep->send_cq)
+				ep->tx_enabled = true;
+			if (ep->recv_cq)
+				ep->rx_enabled = true;
+		}
+
+		break;
+	case FI_GETFIDFLAG:
+	case FI_SETFIDFLAG:
+	case FI_ALIAS:
+	default:
+		return -FI_ENOSYS;
+	}
+err:
+	return ret;
+}
+
+static void __sep_destruct(void *obj)
+{
+	int i;
+	struct fid_domain *domain;
+	struct gnix_fid_ep *ep;
+	struct gnix_fid_domain *domain_priv;
+	struct gnix_fid_sep *sep = (struct gnix_fid_sep *) obj;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	domain = sep->domain;
+	assert(domain != NULL);
+	domain_priv = container_of(domain, struct gnix_fid_domain, domain_fid);
+
+	_gnix_ref_put(domain_priv);
+
+	for (i = 0; i < sep->info->ep_attr->tx_ctx_cnt; i++) {
+		ep = container_of(sep->ep_table[i],
+				  struct gnix_fid_ep, ep_fid);
+		if (ep == NULL) {
+			continue;
+		}
+
+		if (ep->av) {
+			_gnix_ref_put(ep->av);
+		}
+	}
+
+	if (sep->ep_table)
```

Comment:
You should probably put the previous for loop in this conditional also, since it dereferences ep->ep_table also.